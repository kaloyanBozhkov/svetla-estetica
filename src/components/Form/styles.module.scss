.form {
    display: flex;
    flex-direction: column;
    align-items: stretch;

    [form-field] {
        &:not(:last-child) {
            margin-bottom: 2rem;
        }
    }

    button {
        margin: 1.6rem auto;
    }
}
