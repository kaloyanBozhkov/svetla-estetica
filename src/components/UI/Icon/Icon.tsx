import type { ReactElement } from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'

// import icons
import { faSpinner } from '@fortawesome/free-solid-svg-icons'

import styles from './styles.module.scss'

const Icon = ({ icon }: { icon: string }): ReactElement => {
    switch (icon) {
        case 'spinner':
            return <FontAwesomeIcon icon={faSpinner} className={styles.spinner} />

        default:
            return <i>Icon not found</i>
    }
}

export default Icon
