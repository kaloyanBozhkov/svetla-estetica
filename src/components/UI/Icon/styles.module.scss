@import "scss/mixins.scss";

.spinner {
    @include spinAnimation;
}
