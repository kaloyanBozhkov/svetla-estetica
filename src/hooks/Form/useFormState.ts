import type { FormFieldDefinition, FormFieldProps, FormState } from 'common/types'
import { useReducer, Dispatch } from 'react'

type DispatchedChange = {
    fieldName: FormFieldProps['name']
    fieldValue: FormFieldProps['value']
}

// initializes our form state from the definitions passed to Form
export const formStateInit = (formFieldDefinitions: FormFieldDefinition[]): FormState =>
    formFieldDefinitions.reduce(
        (acc, field) => ({
            ...acc,
            [field.name]: field,
        }),
        {}
    )

const useFormState = (
    formFieldDefinitions: FormFieldDefinition[]
): [FormState, Dispatch<DispatchedChange>] => {
    const [formState, dispatchChange] = useReducer<
        (acc: FormState, dispatchedChange: DispatchedChange) => FormState,
        FormFieldDefinition[]
    >(
        (acc, dispatchedChange) => ({
            ...acc,
            // get the changed value for the field name, keep other config as is
            [dispatchedChange.fieldName]: {
                ...acc[dispatchedChange.fieldName],
                value: dispatchedChange.fieldValue,
            },
        }),
        formFieldDefinitions,
        formStateInit
    )

    return [formState, dispatchChange]
}

export default useFormState
