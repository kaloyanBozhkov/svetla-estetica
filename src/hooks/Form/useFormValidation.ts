import { useState } from 'react'
import type { FormState } from 'common/types'

/**
 * Takes in an array of form field definition objects and returns an array of invalid fields that did not pass their .validation()
 * @param formDieldDefinitions
 * @returns [invalidFields, onValidate, onResetValidation] -> onValidate returns true if all fields valid, or false if even one is invalid
 */
const useFormValidation = (
    formState: FormState
): [Record<string, string>, () => boolean, (fieldName?: string) => void] => {
    const [invalidFields, setInvalidFields] = useState<Record<string, string>>({}),
        onValidate = () => {
            // update invalidFields to only hold fieldName of invalid fields
            const invalidFieldsArr = Object.values(formState).reduce((acc, field) => {
                // field is required
                if (field.required) {
                    // field validation returns true if valid or string with err msg if invalid
                    const validation = field.validation(formState)

                    return validation.isValid
                        ? acc
                        : {
                              ...acc,
                              [field.name]:
                                  validation.error || 'Field is invalid - no error msg setup',
                          }
                }

                return acc
            }, {} as Record<string, string>)

            setInvalidFields(invalidFieldsArr)

            // return true if all fields valid, or false if at least one invalid
            return !Object.values(invalidFieldsArr).length
        },
        onResetValidation = (fieldName?: string) =>
            setInvalidFields(
                fieldName
                    ? // if for field provided let's reset just that one
                      Object.keys(invalidFields).reduce(
                          (acc, property) =>
                              fieldName === property
                                  ? acc
                                  : { ...acc, [property]: invalidFields[property] },
                          {}
                      )
                    : // otherwise reset all invalid fields
                      {}
            )

    return [invalidFields, onValidate, onResetValidation]
}

export default useFormValidation
