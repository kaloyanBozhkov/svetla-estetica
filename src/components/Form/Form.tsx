// import common types
import type { ReactElement } from 'react'
import type { FormFieldDefinition, FormState } from 'common/types'
import { InputProps } from 'components/UI/Input/Input'

// import hooks
import useFormState from 'hooks/Form/useFormState'
import useFormValidation from 'hooks/Form/useFormValidation'

// import components
import Button from 'components/UI/Button/Button'
import Icon from 'components/UI/Icon/Icon'
import FormField from './FormField/FormField'

// import styles
import styles from './styles.module.scss'

type FormProps = {
    definitions: FormFieldDefinition[]
    formId?: string
    isSubmitPending?: boolean
    submitLabel: string
    fieldModifier?: InputProps['modifier']
    // BE provided errors obj following submit
    fieldErrors?: Record<string, string>
    onSubmit?: (formState: FormState) => void | boolean
    onFormStateChanged?: (fieldName: string) => void
    onFormFieldFocused?: (fieldName: string) => void
}

/**
 *
 * @param formId string to identify form element and apply custom styles from parent
 * @param action form action attribute
 * @param method 'POST' | 'GET'
 * @param definitions FormFieldDefinition[] -> what fields to render, what their label/validation/requirement is etc..
 * @param isSubmitPending boolean -> renders loading spinner instead of btn indicative or pending operation
 * @param submitLabel string for submit btn label
 * @param fieldErrors
 * @param onSubmit fn to call on submit click, prevents default behavior runs fn and then submits form
 * @param onFormStateChanged fn to call when form state has changed -> useful for events relative to on form field change
 * @param onFormFieldFocused fn to call when a form field has been focused -> useful for removing fieldErrors or triggering smth on form field focus
 * @returns Form ReactElement
 */
const Form = ({
    formId,
    definitions,
    submitLabel = 'Submit',
    fieldModifier = 'base',
    fieldErrors = {},
    isSubmitPending = false,
    onSubmit,
    onFormStateChanged,
    onFormFieldFocused,
}: FormProps): ReactElement => {
    const [formState, dispatchChange] = useFormState(definitions),
        [invalidFields, onValidate, onResetValidation] = useFormValidation(formState)

    return (
        <form className={styles.form} form-id={formId}>
            {Object.values(formState).map((field) => (
                <FormField
                    key={field.name}
                    type={field.type}
                    name={field.name}
                    label={field.label}
                    value={field.value}
                    required={field.required}
                    modifier={fieldModifier}
                    // either from FE's invalidFields (pre submit) or BE's fieldErrors (post submit)
                    errorMsg={invalidFields[field.name] || fieldErrors[field.name]}
                    description={field.description}
                    hasErrored={!!invalidFields[field.name] || !!fieldErrors[field.name]}
                    onChange={({ target: { value: fieldValue } }) => {
                        // update field value
                        dispatchChange({ fieldValue, fieldName: field.name })

                        if (invalidFields[field.name]) {
                            onResetValidation(field.name)
                        }

                        // on change of field if we have a fn to run let's run it and pass field name of changed form field
                        onFormStateChanged?.(field.name)
                    }}
                    onFocus={() => {
                        // if field is invalid let's reset it
                        if (invalidFields[field.name]) {
                            onResetValidation(field.name)
                        }

                        // on focus of field run the fn if we have passed it as prop
                        onFormFieldFocused?.(field.name)
                    }}
                />
            ))}
            {isSubmitPending ? (
                <Button type="button" label={<Icon icon="spinner" />} />
            ) : (
                <Button
                    modifier={(Object.values(invalidFields).length && 'denied') || undefined}
                    type="submit"
                    label={submitLabel}
                    onClick={(e) => {
                        // prevent default form behavior
                        e.preventDefault()

                        if (onValidate()) {
                            onSubmit?.(formState)
                        }
                    }}
                />
            )}
        </form>
    )
}

export default Form
