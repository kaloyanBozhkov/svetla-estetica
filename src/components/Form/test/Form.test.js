import React from 'react'
import { shallow } from 'enzyme'

// import hooks
import * as useFormValidationModuels from 'hooks/Form/useFormValidation'
import * as useFormStateModules from 'hooks/Form/useFormState'

// import atom
import Icon from 'components/UI/Icon/Icon'

// import component
import Form from '../Form'

describe('testing Form', () => {
    const definitions = [
        {
            type: 'text',
            name: 'email',
            label: 'label',
            required: true,
            validation() {
                return this.required && !!this.value
            },
        },
        {
            type: 'password',
            name: 'password',
            label: 'label2',
            required: true,
            validation() {
                return this.required && !!this.value
            },
        },
    ]

    it('should render component', () => {
        const wrapper = shallow(<Form formId="someId" definitions={definitions} />)
        expect(wrapper).toMatchSnapshot()
    })

    it('should pass our field definitions to useFormState hook and also on change of field should dispatch the hooks dispatchChange', () => {
        const mockDispatchChange = jest.fn(),
            mockUseFormState = jest.fn(() => [
                {
                    [definitions[0].name]: {
                        ...definitions[0],
                    },
                },
                mockDispatchChange,
            ])

        jest.spyOn(useFormStateModules, 'default').mockImplementationOnce(mockUseFormState)

        expect(useFormStateModules.default).toHaveBeenCalledTimes(0)

        const wrapper = shallow(<Form formId="someId" definitions={definitions} />)

        // hook called on mount
        expect(useFormStateModules.default).toHaveBeenCalledTimes(1)

        // hook called wiht our definitions
        expect(useFormStateModules.default).toHaveBeenCalledWith(definitions)

        expect(mockDispatchChange).toHaveBeenCalledTimes(0)

        wrapper.find('FormField').prop('onChange')({ target: { value: 'new value' } })

        expect(mockDispatchChange).toHaveBeenCalledTimes(1)

        expect(mockDispatchChange).toHaveBeenCalledWith({
            fieldName: 'email',
            fieldValue: 'new value',
        })
    })

    it('should run onFormStateChanged for FormField when field value changed', () => {
        const mockDispatchChange = jest.fn(),
            mockFormStateChanged = jest.fn(),
            mockUseFormState = jest.fn(() => [
                {
                    [definitions[0].name]: {
                        ...definitions[0],
                    },
                },
                mockDispatchChange,
            ])

        jest.spyOn(useFormStateModules, 'default').mockImplementationOnce(mockUseFormState)

        const wrapper = shallow(
            <Form
                formId="someId"
                definitions={definitions}
                onFormStateChanged={mockFormStateChanged}
            />
        )

        expect(mockFormStateChanged).toHaveBeenCalledTimes(0)

        wrapper.find('FormField').prop('onChange')({ target: { value: 'new value' } })

        expect(mockFormStateChanged).toHaveBeenCalledTimes(1)
    })

    it('should run onFormFieldFocused for FormField when field focused', () => {
        const mockDispatchChange = jest.fn(),
            mockFormFieldFocused = jest.fn(),
            mockUseFormState = jest.fn(() => [
                {
                    [definitions[0].name]: {
                        ...definitions[0],
                    },
                },
                mockDispatchChange,
            ])

        jest.spyOn(useFormStateModules, 'default').mockImplementationOnce(mockUseFormState)

        const wrapper = shallow(
            <Form
                formId="someId"
                definitions={definitions}
                onFormFieldFocused={mockFormFieldFocused}
            />
        )

        expect(mockFormFieldFocused).toHaveBeenCalledTimes(0)

        wrapper.find('FormField').prop('onFocus')()

        expect(mockFormFieldFocused).toHaveBeenCalledTimes(1)
        expect(mockFormFieldFocused).toHaveBeenCalledWith('email')
    })

    it('should run validation fns when submit pressed and then run onSubmit', () => {
        // return false to indicate at least 1 invalid field
        const mockOnValidate = jest.fn(() => false),
            mockOnSubmit = jest.fn(),
            mockPreventDefault = jest.fn(),
            mockUseFormValidation = jest.fn(() => [
                { fieldName: 'invalid error msg' },
                mockOnValidate,
            ]),
            spy = jest
                .spyOn(useFormValidationModuels, 'default')
                .mockImplementation(mockUseFormValidation)

        expect(useFormValidationModuels.default).toHaveBeenCalledTimes(0)

        const wrapper = shallow(
            <Form formId="someId" definitions={definitions} onSubmit={mockOnSubmit} />
        )

        // hook called on mount
        expect(useFormValidationModuels.default).toHaveBeenCalledTimes(1)

        expect(mockOnValidate).toHaveBeenCalledTimes(0)

        // submit form
        wrapper.find('Button').prop('onClick')({ preventDefault: mockPreventDefault })

        // e.preventDefault called and form submit behavior stopped
        expect(mockPreventDefault).toHaveBeenCalledTimes(1)

        // onValidate called
        expect(mockOnValidate).toHaveBeenCalledTimes(1)

        // onSubmit not called
        expect(mockOnSubmit).toHaveBeenCalledTimes(0)

        spy.mockRestore()
    })

    it('should run validation fns when submit pressed but not run onSubmit fn since validaiton returned fale', () => {
        // return true to simulate no invalid fields
        const mockOnValidate = jest.fn(() => true),
            mockOnSubmit = jest.fn(),
            mockPreventDefault = jest.fn(),
            mockUseFormValidation = jest.fn(() => [['invalidFieldName'], mockOnValidate]),
            spy = jest
                .spyOn(useFormValidationModuels, 'default')
                .mockImplementation(mockUseFormValidation)

        expect(useFormValidationModuels.default).toHaveBeenCalledTimes(0)

        const wrapper = shallow(
            <Form formId="someId" definitions={definitions} onSubmit={mockOnSubmit} />
        )

        // hook called on mount
        expect(useFormValidationModuels.default).toHaveBeenCalledTimes(1)

        expect(mockOnValidate).toHaveBeenCalledTimes(0)

        // submit form
        wrapper.find('Button').prop('onClick')({ preventDefault: mockPreventDefault })

        // e.preventDefault called and form submit behavior stopped
        expect(mockPreventDefault).toHaveBeenCalledTimes(1)

        // onValidate called
        expect(mockOnValidate).toHaveBeenCalledTimes(1)

        // onSubmit called
        expect(mockOnSubmit).toHaveBeenCalledTimes(1)

        spy.mockReset()
    })

    it('should set hasErrored to FormField if invalidInputs obj has the field name', () => {
        const mockOnValidate = jest.fn(() => 'set invalid field name'),
            mockDispatchChange = jest.fn(),
            mockOnResetValidation = jest.fn(),
            mockUseFormValidation = jest.fn(() => [
                { [definitions[0].name]: 'error msg' },
                mockOnValidate,
                mockOnResetValidation,
            ]),
            spy = jest
                .spyOn(useFormValidationModuels, 'default')
                .mockImplementation(mockUseFormValidation),
            mockUseFormState = jest.fn(() => [
                {
                    [definitions[0].name]: {
                        ...definitions[0],
                    },
                    [definitions[1].name]: {
                        ...definitions[1],
                    },
                },
                mockDispatchChange,
            ])

        jest.spyOn(useFormStateModules, 'default').mockImplementationOnce(mockUseFormState)

        const wrapper = shallow(<Form formId="someId" definitions={definitions} />)

        // first field hasError set to true de to
        expect(wrapper.find('FormField').at(0).prop('hasErrored')).toEqual(true)

        // second field does not have it set since useFormValidation hook's returned invalid array does not list the filed name
        expect(wrapper.find('FormField').at(1).prop('hasErrored')).toEqual(false)

        spy.mockReset()
    })

    it('should run onResetValidation with field name if invalidInputs obj has error for that field (on change and focus)', () => {
        const mockOnValidate = jest.fn(() => 'set invalid field name'),
            mockDispatchChange = jest.fn(),
            mockOnResetValidation = jest.fn(),
            mockUseFormValidation = jest.fn(() => [
                { [definitions[0].name]: 'error msg' },
                mockOnValidate,
                mockOnResetValidation,
            ]),
            spy = jest
                .spyOn(useFormValidationModuels, 'default')
                .mockImplementation(mockUseFormValidation),
            mockUseFormState = jest.fn(() => [
                {
                    [definitions[0].name]: {
                        ...definitions[0],
                    },
                    [definitions[1].name]: {
                        ...definitions[1],
                    },
                },
                mockDispatchChange,
            ])

        jest.spyOn(useFormStateModules, 'default').mockImplementationOnce(mockUseFormState)

        const wrapper = shallow(<Form formId="someId" definitions={definitions} />),
            emailFormField = wrapper.find('FormField').at(0)

        expect(mockOnResetValidation).toHaveBeenCalledTimes(0)

        emailFormField.prop('onChange')({ target: { value: 'new value' } })

        expect(mockOnResetValidation).toHaveBeenCalledTimes(1)
        expect(mockOnResetValidation).toHaveBeenLastCalledWith('email')

        emailFormField.prop('onFocus')()

        expect(mockOnResetValidation).toHaveBeenCalledTimes(2)
        expect(mockOnResetValidation).toHaveBeenLastCalledWith('email')

        spy.mockReset()
    })

    it('should have a submit btn with modifier set to denied if invalidFields contain a value', () => {
        const mockOnValidate = jest.fn(() => 'set invalid field name'),
            mockDispatchChange = jest.fn(),
            mockOnResetValidation = jest.fn(),
            mockUseFormValidation = jest.fn(() => [
                { [definitions[0].name]: 'error msg' },
                mockOnValidate,
                mockOnResetValidation,
            ]),
            spy = jest
                .spyOn(useFormValidationModuels, 'default')
                .mockImplementation(mockUseFormValidation),
            mockUseFormState = jest.fn(() => [
                {
                    [definitions[0].name]: {
                        ...definitions[0],
                    },
                },
                mockDispatchChange,
            ])

        jest.spyOn(useFormStateModules, 'default').mockImplementationOnce(mockUseFormState)

        const wrapper = shallow(<Form formId="someId" definitions={definitions} />)

        // first field hasError set to true de to
        expect(wrapper.find('Button').prop('modifier')).toEqual('denied')

        spy.mockReset()
    })

    it('should render button with spinner if isSubmitPending set to true', () => {
        const mockOnValidate = jest.fn(() => 'set invalid field name'),
            mockDispatchChange = jest.fn(),
            mockOnResetValidation = jest.fn(),
            mockUseFormValidation = jest.fn(() => [
                { [definitions[0].name]: 'error msg' },
                mockOnValidate,
                mockOnResetValidation,
            ]),
            spy = jest
                .spyOn(useFormValidationModuels, 'default')
                .mockImplementation(mockUseFormValidation),
            mockUseFormState = jest.fn(() => [
                {
                    [definitions[0].name]: {
                        ...definitions[0],
                    },
                },
                mockDispatchChange,
            ])

        jest.spyOn(useFormStateModules, 'default').mockImplementationOnce(mockUseFormState)

        const wrapper = shallow(<Form formId="someId" definitions={definitions} isSubmitPending />)

        // btn has rendered an icon spinner instead of a label
        expect(wrapper.find('Button').prop('label')).toEqual(<Icon icon="spinner" />)

        spy.mockReset()
    })
})
