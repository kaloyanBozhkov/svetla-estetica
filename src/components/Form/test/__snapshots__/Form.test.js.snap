// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing Form should render component 1`] = `
<form
  className="form"
  form-id="someId"
>
  <FormField
    hasErrored={false}
    key="email"
    label="label"
    modifier="base"
    name="email"
    onChange={[Function]}
    onFocus={[Function]}
    required={true}
    type="text"
  />
  <FormField
    hasErrored={false}
    key="password"
    label="label2"
    modifier="base"
    name="password"
    onChange={[Function]}
    onFocus={[Function]}
    required={true}
    type="password"
  />
  <Button
    label="Submit"
    onClick={[Function]}
    type="submit"
  />
</form>
`;
